Add support for F3

aarshkshah1992 opened the pull request:
* Add support for libp2p and  HTTP over libp2p transports
* Script that allows to run and plot time_to_fetch for any composition.

### TODO
- [ ] Support Concurrency.
- [ ] Support config for Libp2p Muxer and Security.

aarshkshah1992 committed:
add support for F3

aarshkshah1992 committed:
remove http for now

aarshkshah1992 committed:
add HTTP transport

aarshkshah1992 committed:
comment HTTP

aarshkshah1992 committed:
measure gs time correctly

aarshkshah1992 commented on testbed/testbed/utils/graphsync.go:
```diff
@@ -109,9 +111,13 @@ func (n *GraphsyncNode) Fetch(ctx context.Context, c cid.Cid, peers []PeerInfo)
 	resps, errs := n.gs.Request(ctx, p, cidlink.Link{Cid: c}, selectAll)
 	for range resps {
 	}
+	fmt.Println("TIME SINCE START: ", time.Since(start))
```
@hannahhoward @adlrocha 

I think this is the correct time it took for the GS request, right ? I see a difference of almost 20 seconds for a 1G file if we don't measure it here and measure it after the call returns.

Reasons:

1. The `n.gs.Request` call already writes the blocks to the datastore.
2. We do one more read in the `n.dserv.Get(ctx, c)` call below (I think ?)
3. In the transfer code, we call `err = files.WriteTo(rcvFile, "/tmp/"+strconv.Itoa(t.tpindex)+time.Now().String())` and do one more write of the whole file. 

So, we should ONLY measure 1. Wdyt ?

adlrocha replied:
@hannahhoward is definitely the expert here. I recall that in Bitswap we did it like this because if you don't explicitly write the file, UnixFS doesn't know if there are blocks pending, and Bitswap doesn't traverse the full DAG. I don't think this is the case for GS, but can't be sure.

hannahhoward replied:
This is an interesting result. After you've read all the responses, the file should indeed be transferred. 

`n.dserve.Get(ctx, c)` actually only reads the root block from the blockstore, but `err = files.WriteTo(rcvFile, "/tmp/"+strconv.Itoa(t.tpindex)+time.Now().String())` will trigger reading the rest of the file from the blockstore as it writes out to a flat file.

We had the measurement measuring after the file was written, because we wanted to measure the overall time to get the file from a user perspective, which is a bit different than here, where you're measuring time to get the file from a pure transfer perspective.

That all said, I'm honestly quite surprised it's 20 seconds more, since it should be all local i/o. Then again, you're getting into unixfs code and there could be inefficiencies. It's an interesting result not for your measurements, but for general IPFS performance. I'd like to maybe seperate that timing in the reporting for all modes of transfer, if that's possible. I may submit a PR to suggest a direction here.

hannahhoward replied:
ha! I know what this is about and it indicates why you make be getting slow results for Graphsync transfer as well. There is a 5 millisecond delay on blockstore reads by default! For a 1GB file, at default chunking settings for this test (256K), that's >4000 blocks x 5 milliseconds = 20 seconds!

You are going to want to set bstore_delay_ms = 0 in your composition for a fair comparison with any of your other protocols.

aarshkshah1992 replied:
I've added support for using a Badger data store in addition to the mutex wrapped in-memory store with an artificial delay. Will get back with results using:

1. Badger
2. BlockStore Delay of 0.
3. BlockStore Delay of 5ms.

hannahhoward replied:
Now that the Blockstore delay issue is clairfied, can we remove these lines?

adlrocha commented on testbed/compositions/graphsync.toml:
```diff
@@ -16,23 +16,20 @@ instances = { count = 2 }
 # file_size = "10000000,30000000,50000000"
-latency_ms= "100,10"
-bandwidth_mb= "100,50"
+latency_ms= "0"
```
Small nitpick.
I guess you are removing latency and bandwidth because you are interested in the performance of graphsync within the peer, right? If this is not the case and you want to test the protocol's performance we should add some of this. 

I would maybe keep latency and bandwidth to the standard 100, 100 or something like that (even if you change it locally for your tests) so we don't encourage zero latency/bandwidth tests.

aarshkshah1992 replied:
@adlrocha I agree. I've reverted the changes in all the composition files. Will explicitly name the files as `no_latency...` to make sure the reader knows the composition does NOT use latency and is ONLY for benchmarking purposes.

adlrocha replied:
Great idea!

adlrocha commented on testbed/compositions/libp2pHTTP.toml:
```diff
@@ -0,0 +1,36 @@
+enable_tcp= "false"
+enable_dht= "false"
+node_type = "libp2pHTTP"
+long_lasting = "true"
```
First time I've seen this new line message in Github haha

adlrocha commented on testbed/compositions/graphsync.toml:
```diff
@@ -16,23 +16,20 @@ instances = { count = 2 }
-input_data = "files"
-data_dir = "../extra/test-datasets"
+input_data = "random"
+file_size = "2000000000"
```
Generating all this data may be really slow, what about using real data?

aarshkshah1992 replied:
@adlrocha @hannahhoward 

I agree with Alfonso here and am putting code to use real files on disk.

One more interesting thing I realised today is that Random File Generator uses an in-memory buffer: 

```
// RandReader generates random data from seed.
func SeededRandReader(len int, seed int64) io.Reader {
	randReader := rand.New(rand.NewSource(seed))
	data := make([]byte, len)
	randReader.Read(data)
	return bytes.NewReader(data)
}
```

and so if we use this, we don't actually do any disk IO on the send side of the transfer.

Opposed to this, when we give a file on disk to the tests, it will cause disk I/O and will be more in-line with GS using Badger. So, we should ONLY use pre-generated files(`PathFile`) going ahead for our benchmarks.

adlrocha commented on testbed/testbed/scripts/pdf_composition.py:
```diff
@@ -0,0 +1,24 @@
+
+print(filename)
+
+with PdfPages(dir_path + "/../../compositions/results/"+filename) as export_pdf:
```
Why not adding this to `pdf.py` instead of having two different scripts?

aarshkshah1992 replied:
Do you mean refactor the pdf.py file to take the paths as args ?

adlrocha replied:
Yep! Instead of receiving just the filename, add the absolute path. That way we don't have different scripts with a lot of redundant code.

adlrocha commented on testbed/testbed/utils/libp2pHTTP.go:
```diff
@@ -0,0 +1,139 @@
+	return nil
+}
+
+func randCid() (cid.Cid, error) {
```
Out of a curiosity, why are you using a random CID instead of using the CID for the file? 🙈

hannahhoward replied:
Since there's no import to a blockstore, there's no IPLD graph generated, and therefor no obvious CID generated.

However, I think the right thing to do here, and it's a relatively small change, is to multihash sum all the bytes in the file, instead of a random integer. That will mean that the CID is in fact the correct CID for a raw block that contains all the bytes of the file, and more importantly will make it deterministic based on the input of the file.

The downside of random CID is if at some point we want to use more than one seed, we're going to end up with each seed generating different CIDs. The suggested approach of summing all the bytes in the file should enable all seeds to generate the same CID.

aarshkshah1992 replied:
@hannahhoward @adlrocha 

The problem here is that I'd have to read the file to get it's hash. But, we need to read it again during sending it to the user.  I tried using the `Seek` method on the `files.File` interface to reset the offset after hashing it but looks like it isn't always supported.

Is there a clean way of reading this same `files.File` twice ? Copying an interface ONLY copies the interface header and NOT the underlying Reader implementation here.

hannahhoward replied:
You could certainly use an io.TeeReader to write into a secondary temporary file as you hash and serve that.

hannahhoward replied:
non-blockign for this PR I think -- just something you're going to need to do to support more than one seed.

adlrocha commented on testbed/testbed/utils/rawLibp2p.go:
```diff
@@ -0,0 +1,95 @@
+	h host.Host
+}
+
+func CreateRawLibp2pNode(ctx context.Context, h host.Host, nodeTP NodeType) (*RawLibp2pNode, error) {
```
What about merging this code with the one for the HTTPlibp2pNode so we have a single libp2pNode definition where we register handlers, and an attribute in the libp2p node that defines the type of exchange to perform? That way we can easily accommodate new protocols over the libp2p node without a lot of code duplication. What do you think?

hannahhoward replied:
honestly they look quite a bit different. I feel like part of the node interface's point it to be the container for more protocol tests --I'm personally ok with these being two different node implementations.

hannahhoward replied:
but I also see your point.

aarshkshah1992 replied:
Going ahead with different files for now as the implementation does differ considerably.

adlrocha replied:
After a second review, I am convinced :) Go for it!

aarshkshah1992 commented on testbed/compositions/graphsync.toml:
```diff
@@ -16,23 +16,20 @@ instances = { count = 2 }
-input_data = "files"
-data_dir = "../extra/test-datasets"
+input_data = "random"
+file_size = "2000000000"
```
@hannahhoward This dies when I try to randomly generate and add a 4GB file in Graphsync Node. It works for other nodes though. Any idea why this is ?

aarshkshah1992 replied:
I think probably better to use predefined files here for repeatable tests.

hannahhoward replied:
This works for smaller numbers. Import via graphsync puts the data into a blockstore, which is all in memory currently, using IPFS's adding code (see DAGAdder). It doesn't totally surprisme me it breaks for a file this large. I would recommend adding disk support (i.e. badger) which is not currently supported in this repo I think.

hannahhoward replied:
also, I would recommend making new composition files for your parameters for what you want to test. These composition files may be in use by other teams. I recommend making files like:
no_latency_protocol_comparison_graphsync.toml
no_latency_protocol_comparison_pure_http.toml
no_latency_protocol_comparison_http_over_libp2p.toml
etc.

hannahhoward replied:
this will enable the ignite team to customize your benchmarks without overwritng other teams benchmarks.

aarshkshah1992 replied:
@hannahhoward

1. Added Badger support to this repo so we can test with large files.
2. Reverted all changes to composition files. Will use the naming scheme you mention so as to not step on toes going ahead.

hannahhoward commented on testbed/testbed/utils/libp2pHTTP.go:
```diff
@@ -0,0 +1,139 @@
+		// start an http server on port 8080
+		svr := &http.Server{}
+		go svr.Serve(listener)
+		time.Sleep(1 * time.Second)
```
what's with the sleep here?

aarshkshah1992 replied:
To ensure the server is up. Dosen't hurt I think.

hannahhoward replied:
for sure, and that's fine here.

As a point of consideration, not for needing to change, but more food for thought:

When I see a time.Sleep, I tend to want to ask why--
-- what is it I think I need to be waiting for?
-- do I actually need to wait?
-- what does the documentation of the function say about needing to wait?
-- Is there a more deterministic way to wait -- i.e. will there be an asynchronous return over a channel I can read? Can I set this up somehow?

hannahhoward reviewed (changes requested):
Amazing this all got done in a single evening! Awesome work.

I'd prefer we make seperate compositions for the tests that Ignite is going to run if possible, since by and large you all test with no latency, will want to turn the block store delay down to 0, etc -- and the data systems team is going to be using this repo actively for regression testing. Having separate compositions will keep us from overriding each others work.

I suggested a naming scheme in one of my comments, but feel free to invent your own.

All the other comments are non-blocking, but you should definitely note the comment about blockstore delay to make your comparison accurate.

aarshkshah1992 committed:
add badger support and fix composition files

aarshkshah1992 committed:
fix composition

aarshkshah1992 commented on testbed/testbed/go.mod:
```diff
@@ -32,6 +33,8 @@ require (
 	github.com/libp2p/go-libp2p v0.11.0
 	github.com/libp2p/go-libp2p-core v0.6.1
+	github.com/libp2p/go-libp2p-gostream v0.2.1
+	github.com/libp2p/go-libp2p-http v0.1.6-0.20210310045043-5508c68db693
```
This commit configures the HTTP client to use the same buffer size as Graphsync (4 mb) while reading from the sockets.

aarshkshah1992 commented:
@adlrocha @hannahhoward 

1. Added support for Badger to the repo.
2. Reverted changes on the existing composition files.
3. Added a [script](https://github.com/protocol/beyond-bitswap/pull/41/commits/82de5ddc85589bfd04ea4221546f1c70569ad139) to generate files on disk to use for F3 (and maybe others).

I've addressed some of the changes and have some questions regarding implementing the other ones. Please can you take a look at my comments ?

aarshkshah1992 committed:
more changes

aarshkshah1992 committed:
script to generate files

aarshkshah1992 committed:
fixed error handling in file script

adlrocha commented on testbed/test-datasets/f3/gen_file.go:
```diff
@@ -0,0 +1,39 @@
+package main
```
Quick note. There's already a scripts/random-file.sh in the repo in case you want to generate a bunch of random files, add them to the dataset directory, and use them for every test so you don't have to explicitly generate them again.

But this LGTM if you rather generate them with go instead of using the bash script :)

aarshkshah1992 replied:
@adlrocha Based on an offline discussion with @raulk , turns out generating files and them copying them into the docker container while running the tests slows down the execution of the test. Can I go ahead and change the random file generator to create files on disk so they are created within the container itself ?

adlrocha replied:
Go for it!

adlrocha commented:
Awesome work @aarshkshah1992!! Overall everything LGTM! Looking forward to seeing your results :) And thank you @hannahhoward for your support reviewing this.

aarshkshah1992 committed:
random files should be written to disk

aarshkshah1992 committed:
fix TCP

aarshkshah1992 committed:
badger deps

hannahhoward commented on testbed/testbed/.gitignore:
```diff
@@ -2,4 +2,6 @@ scripts/env
\ No newline at end of file
+scripts/__pycache__
+.DS_Store
+../test-datasets/*
```
doesn't this need to be in the root .gitignore?

hannahhoward commented on testbed/testbed/utils/bitswap.go:
```diff
@@ -52,14 +57,44 @@ func (n *BitswapNode) Close() error {
+		return nil, err
+	}
+
+	return datastore, nil
```
Should we still wrap the data store in a block store delay? Otherwise the parameter has no effect when disk store is on. I feel like that should at least be documented.

hannahhoward reviewed (approved):
I put a bunch of comments but overall I'm fine if you want to merge and defer further changes to future PRs.
